### PR TITLE
Filler update

### PR DESCRIPTION
**Pull request by phiedw**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
The filler update function has not been implemented

**What is the new behavior (if this is a feature change)?**
The filler update function now updates the reference flows, the sensitivities, and the range action negative and positive max variations according to a new systematic sensitivity analysis result (for the flows and the sensitivities) and a list of remedial action results (for the new range action variations)

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/CoreProblemFiller.java [hunk 1]
```diff
@@ -86,4 +107,28 @@ private void updateCnecConstraintWithRangeAction(Cnec cnec, RangeAction rangeAct
...
+
+    private double getRemedialActionResultValue(RemedialActionElementResult remedialActionElementResult) {
+        if (remedialActionElementResult instanceof PstElementResult) {
+            PstElementResult pstElementResult = (PstElementResult) remedialActionElementResult;
+            return pstElementResult.getPostOptimisationAngle() - pstElementResult.getPreOptimisationAngle();
+        } else if (remedialActionElementResult instanceof RedispatchElementResult) {
+            RedispatchElementResult redispatchElementResult = (RedispatchElementResult) remedialActionElementResult;
+            return redispatchElementResult.getPostOptimisationTargetP() - redispatchElementResult.getPreOptimisationTargetP();
+        }
+        return 0;
```
Comment (terriervik) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/CoreProblemFiller.java]: To do...

Comment (terriervik) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/CoreProblemFiller.java]: Please return an error instead.

Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/CoreProblemFiller.java]: ok

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java [hunk 2]
```diff
@@ -120,21 +122,80 @@ private double getRemedialActionResultValue(RemedialActionResult remedialActionR
         throw new FaraoException("Range action type of " + remedialActionElementResult.getId() + " is not supported yet");
     }
 
-    private boolean sameRemedialActionResultLists(List<RemedialActionResult> firstList, List<RemedialActionResult> secondList) {
-        if (firstList.size() != secondList.size()) {
-            return false;
+    private RemedialActionResult combineRAResults(RemedialActionResult oldRemedialActionResult, RemedialActionResult newRemedialActionResult) {
+        //mapping the old RA elements to their IDs
+        Map<String, RemedialActionElementResult> oldRAResultElementMap = new HashMap<>();
```
Comment (TheMaskedTurtle) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: you can make it inline like that `Map<String, RemedialActionElementResult> oldRAResultElementMap = oldRemedialActionResult.getRemedialActionElementResults().stream()
            .collect(Collectors.toMap(RemedialActionElementResult::getId, remedialActionElementResult -> remedialActionElementResult));`
It is more java 8-ish

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java [hunk 3]
```diff
@@ -120,21 +122,80 @@ private double getRemedialActionResultValue(RemedialActionResult remedialActionR
...
-                    Math.abs(firstMap.get(remedialActionResult.getId()) - getRemedialActionResultValue(remedialActionResult)) > 0.001) {
-                return false;
+        for (RemedialActionResult remedialActionResult : remedialActionResultHistoryList) {
+            if (!newRAResultMap.containsKey(remedialActionResult.getId())) {
+                newRemedialActionResultHistoryList.add(remedialActionResult);
             }
         }
-        return true;
+        remedialActionResultHistoryList = newRemedialActionResultHistoryList;
     }
```
Comment (TheMaskedTurtle) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: Can be hugely reduced by defining `remedialActionResultHistoryList` as a Map directly

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/AbstractProblemFiller.java [hunk 4]
```diff
@@ -21,7 +25,7 @@ public AbstractProblemFiller(LinearRaoProblem linearRaoProblem, LinearRaoData li
 
     public abstract void fill();
 
-    public void update(LinearRaoProblem linearRaoProblem, LinearRaoData linearRaoData) {
+    public void update(LinearRaoProblem linearRaoProblem, LinearRaoData linearRaoData, List<RemedialActionResult> remedialActionResultList) {
```
Comment (TheMaskedTurtle) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/AbstractProblemFiller.java]: I would have removed the todo, declared the method as abstract, and implement it as empty in the positiveMinMarginFiller

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java [hunk 5]
```diff
@@ -56,17 +57,19 @@ public String getVersion() {
         preOptimSensitivityAnalysisResult = SystematicSensitivityAnalysisService.runAnalysis(network, crac, computationManager);
         postOptimSensitivityAnalysisResult = preOptimSensitivityAnalysisResult;
         SystematicSensitivityAnalysisResult midOptimSensitivityAnalysisResult;
```
Comment (TheMaskedTurtle) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: can be renamed tempSensitivityAnalysisResult

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java [hunk 6]
```diff
@@ -120,21 +122,80 @@ private double getRemedialActionResultValue(RemedialActionResult remedialActionR
...
+        /*
+        For each RA element of the second RA, we create a new RA element combining that RA element and the corresponding
+          one from the first RA (as long as they dont cancel each other out)
+        */
+        for (RemedialActionElementResult remedialActionElementResult : newRemedialActionResult.getRemedialActionElementResults()) {
+            if (remedialActionElementResult instanceof PstElementResult) {
+                PstElementResult newPstElementResult = (PstElementResult) remedialActionElementResult;
+                PstElementResult oldPstElementResult = (PstElementResult) oldRAResultElementMap.get(remedialActionElementResult.getId());
+                double totalChange = newPstElementResult.getPostOptimisationAngle() - oldPstElementResult.getPreOptimisationAngle();
+                if (totalChange > 0.0001) {
```
Comment (TheMaskedTurtle) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: You can define a static final variable to define this minimum threshold

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/CoreProblemFiller.java [hunk 7]
```diff
@@ -91,4 +113,28 @@ private void updateCnecConstraintWithRangeAction(Cnec cnec, RangeAction rangeAct
                         networkElement.getId(),
                         linearRaoData.getSensitivity(cnec, rangeAction)));
     }
+
+    private double getRemedialActionResultValue(RemedialActionElementResult remedialActionElementResult) {
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/CoreProblemFiller.java]: It returns the post optimisation value.
Name changed to be more clear.

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java [hunk 8]
```diff
@@ -120,21 +122,80 @@ private double getRemedialActionResultValue(RemedialActionResult remedialActionR
...
+        List<RemedialActionElementResult> combinedRAResultElementList = new ArrayList<>();
+        /*
+        For each RA element of the second RA, we create a new RA element combining that RA element and the corresponding
+          one from the first RA (as long as they dont cancel each other out)
+        */
+        for (RemedialActionElementResult remedialActionElementResult : newRemedialActionResult.getRemedialActionElementResults()) {
+            if (remedialActionElementResult instanceof PstElementResult) {
+                PstElementResult newPstElementResult = (PstElementResult) remedialActionElementResult;
+                PstElementResult oldPstElementResult = (PstElementResult) oldRAResultElementMap.get(remedialActionElementResult.getId());
+                double totalChange = newPstElementResult.getPostOptimisationAngle() - oldPstElementResult.getPreOptimisationAngle();
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: indeed

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java [hunk 9]
```diff
@@ -120,21 +122,80 @@ private double getRemedialActionResultValue(RemedialActionResult remedialActionR
...
+                    combinedRAResultElementList.add(new PstElementResult(oldPstElementResult.getId(),
+                            oldPstElementResult.getPreOptimisationAngle(),
+                            oldPstElementResult.getPreOptimisationTapPosition(),
+                            newPstElementResult.getPostOptimisationAngle(),
+                            newPstElementResult.getPostOptimisationTapPosition()));
+                }
+            } else if (remedialActionElementResult instanceof RedispatchElementResult) {
+                RedispatchElementResult newRedispatchElementResult = (RedispatchElementResult) remedialActionElementResult;
+                RedispatchElementResult oldRedispatchElementResult = (RedispatchElementResult) oldRAResultElementMap.get(remedialActionElementResult.getId());
+                double totalChange = newRedispatchElementResult.getPostOptimisationTargetP() - oldRedispatchElementResult.getPreOptimisationTargetP();
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: indeed

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java (hunk 2 above)
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: thanks

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java (hunk 3 above)
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: true

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java (hunk 6 above)
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: ok

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java (hunk 5 above)
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: ok

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/AbstractProblemFiller.java (hunk 4 above)
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/AbstractProblemFiller.java]: ok

**Pull request merged**